File: app/src/main/java/com/ed/redditapp/ui/postlist/Post.java
package com.ed.redditapp.ui.postlist;

import org.json.JSONException;
import org.json.JSONObject;

import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
public class Post {
    private String title;
    private String username;
    private String subreddit;
    private String subredditIconUrl;
    private String permalink;
    private PostThumbnail thumbnailSmall;
    private PostThumbnail thumbnail320;
    private PostThumbnail thumbnail640;
    private PostThumbnail thumbnail960;
    private long timestamp;
    private int commentsCount;
    private int points;

    public Post() {
    }

    public Post(JSONObject json) throws JSONException {
        title = json.getString("title");
        username = json.getString("author");
        subreddit = json.getString("subreddit");
        permalink = json.getString("permalink");
        timestamp = json.getLong("created_utc");
        commentsCount = json.getInt("num_comments");
        points = json.getInt("ups");

        if (json.has("previews")) {
            JSONObject previews = json.getJSONObject("previews");
            if (previews.has("source"))
                thumbnailSmall = new PostThumbnail(previews.getJSONObject("source"));
            if (previews.has("preview320"))
                thumbnail320 = new PostThumbnail(previews.getJSONObject("preview320"));
            if (previews.has("preview640"))
                thumbnail640 = new PostThumbnail(previews.getJSONObject("preview640"));
            if (previews.has("preview960"))
                thumbnail960 = new PostThumbnail(previews.getJSONObject("preview960"));
        }
    }

    public PostThumbnail getLargestThumbnail() {
        if (thumbnail960 != null) return thumbnail960;
        if (thumbnail640 != null) return thumbnail640;
        if (thumbnail320 != null) return thumbnail320;
        if (thumbnailSmall != null) return thumbnailSmall;
        return null;
    }
}

File: app/src/main/java/com/ed/redditapp/lib/api/Post.kt
package com.ed.redditapp.lib.api

abstract class Post {
    abstract val after: String
    abstract val title: String
    abstract val username: String
    abstract val subreddit: String
    abstract val permalink: String
    abstract val domain: String?
    abstract val thumbnailSource: PostThumbnail?
    abstract val thumbnail320: PostThumbnail?
    abstract val thumbnail640: PostThumbnail?
    abstract val thumbnail960: PostThumbnail?
    abstract val timestamp: Long
    abstract val commentsCount: Int
    abstract val points: Int

    abstract val contentType: PostContentType?
    abstract val content: String?

    var subredditIconUrl: String? = null
    val largestThumbnail: PostThumbnail?
        get() {
            if (thumbnail960 != null) return thumbnail960
            if (thumbnail640 != null) return thumbnail640
            if (thumbnail320 != null) return thumbnail320
            return if (thumbnailSource != null) thumbnailSource else null
        }
}
